src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,401 @@
+         * This supports the use-case where a label can be applied to the salt but the actual value
+         * of the salt is not yet available.
+         *
+         * @param salt
+         *     the salt value (null or empty values will not be added)
+         *
+         * @return a new {@code Extract} object
+         */
+        public Builder addSalt(byte[] salt) {
+            if (salt != null && salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "RAW"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Static helper-method that may be used to initialize a {@code Builder} with an empty
+     * {@code Extract}
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should be called next,
+     * before calling build methods, such as {@code Builder.extractOnly()}
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder extract() {
+        return new Builder().createExtract();
+    }
+
+    /**
+     * Static helper-method that may be used to initialize an {@code Expand} object
+     *
+     * @param prk
+     *     the PRK (may be null)
+     * @param info
+     *     the info (may be null)
+     * @param length
+     *     the length
+     *
+     * @return a new {@code Expand} object
+     */
+    static Expand expand(SecretKey prk, byte[] info, int length) {
+        return new Expand(prk, info, length);
+    }
+
+    /**
+     * Static helper-method that may be used to initialize an {@code ExtractExpand} object
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should be called on the
+     * {@code Extract} parameter, before calling this method, since {@code ExtractExpand} is
+     * immutable
+     *
+     * @param ext
+     *     a pre-generated {@code Extract}
+     * @param info
+     *     the info (may be {@code null})
+     * @param length
+     *     the length
+     *
+     * @return a new {@code ExtractExpand} object
+     */
+    static ExtractExpand extractExpand(Extract ext, byte[] info, int length) {
+        return new ExtractExpand(ext, info, length);
+    }
+
+    /**
+     * Defines the input parameters of an Extract operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
+     */
+    class Extract implements HKDFParameterSpec {
+
+        // HKDF-Extract(salt, IKM) -> PRK
+        private final List<SecretKey> ikms;
+        private final List<SecretKey> salts;
+
+        private Extract() {
+            this(new ArrayList<>(), new ArrayList<>());
+        }
+
+        private Extract(List<SecretKey> ikms, List<SecretKey> salts) {
+            this.ikms = ikms;
+            this.salts = salts;
+        }
+
+        /**
+         * Gets the unmodifiable {@code List} of IKM values
+         *
+         * @return the unmodifiable {@code List} of IKM values
+         */
+        public List<SecretKey> ikms() {
```

Comment:
The `ikms` is already unmodifiable when this object is created back in `Builder.extractOnly`. Or, you may move the `copyOf` methods from that method to the constructor in this class.